utils/s2n_init.c

```diff
@@ -105,12 +114,15 @@ int s2n_cleanup(void)
     /* If this is the main thread and atexit cleanup is disabled,
      * perform final cleanup now */
     if (pthread_equal(pthread_self(), main_thread) && !atexit_cleanup) {
+        /* some cleanups are not idempotent (rand_cleanup, for example) so protect */
+        POSIX_ENSURE(initialized, S2N_ERR_NOT_INITIALIZED);
```

Comment:
Sicne we call rand_cleanup before this point, should this check be at the beginning of the s2n_cleanup method?

And if we need to do it for atexit, shouldn't it be in s2n_cleanup_atexit_impl?

Reply:
rand_cleanup is not called before this point; rand_cleanup_thread is; we should not check this at the beginning of the routine because the guidance says to call s2n_cleanup from every thread.  We only want to do this on the last thread out, which is supposed to be the main thread based on the logic here.  Reference counting would be dangerous as I doubt every implementation actually calls s2n_cleanup from every thread, however that would force consumers to do it, seems like it would cause lots of issues so avoiding it.

This initialized flag check does not need to be done in the atexit impl because the atexit handler is only installed in init.  The flag check only matters when atexit_cleanup is false.

As such, I believe the code in the PR is correct and needs to change here.